src/main/scala/org/apache/spark/metrics/source/custom/AccumulatorMetricsSource.scala

```diff
@@ -0,0 +1,54 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.apache.spark.metrics.source.custom
+
+import java.{lang => jl}
+
+import com.codahale.metrics.{Gauge, Meter, MetricRegistry}
+import org.apache.spark.metrics.source.Source
+import org.apache.spark.util.AccumulatorV2
+
+import scala.collection.mutable
+
+/**
+  * Accumulator-based metric source
+  * <p><p>
+  * Usage:
+  * <pre>{@code
+  * val metricsSource = new AccumulatorMetricsSource("prefix")
+  * val metric = spark.sparkContext.longAccumulator("metric-name")
+  * metricsSource.registerAccumulator(metric, metric.name.getOrElse("UNNAMED_ACCUMULATOR"))
+  * SparkEnv.get.metricsSystem.registerSource(metricsSource)
+  * }</pre>
+  * @param prefix for metrics collected by this source
+  */
+class AccumulatorMetricsSource(prefix: String) extends Source {
+  override def sourceName: String = prefix
+  override def metricRegistry: MetricRegistry = registry
+
+  private val registry = new MetricRegistry
+  private val gauges = mutable.Map[String, Long]()
+  private val meters = mutable.Map[String, Meter]()
+
+  def registerAccumulator(acc: AccumulatorV2[jl.Long, jl.Long], name: String) {
+    gauges.put(name, 0L)
+    meters.put(name, metricRegistry.meter(name + "-rate"))
+
+    metricRegistry.register(name,
+      new Gauge[Long] {
+        override def getValue: Long = {
+          val value = acc.value
+          meters(name).mark(getCountOverflowSafe(value, gauges(name)))
```

Comment:
This feels like it's abusing the intent of `getValue` somewhat. Ideally, this function should be simply accessing a precomputed value, but we're including a side effect of updating the meter.

Is it possible for us to provide a mark method that jobs would call directly? Or is the idea here that we need to keep the metrics themselves on the master node so that they don't get serialized and copied on workers?

I certainly don't want us to spin up a thread executor to handle periodically marking the meter, so maybe this is the best we can do.